r/src/altrep.cpp

```diff
@@ -43,144 +50,356 @@ extern "C" {
 
-  static const std::shared_ptr<Array>& Get(SEXP vec) {
-    return *Pointer(R_altrep_data1(vec));
-  }
+  SEXP data2() { return R_altrep_data2(alt_); }
 
-  static R_xlen_t Length(SEXP vec) { return Get(vec)->length(); }
+  R_xlen_t Length() { return array_->length(); }
 
-  static const void* Dataptr_or_null(SEXP vec) {
-    return Get(vec)->data()->template GetValues<data_type>(1);
-  }
+  bool IsMaterialized() { return !Rf_isNull(data2()); }
+
+  void Materialize() {
+    if (!IsMaterialized()) {
+      auto size = array_->length();
 
-  static SEXP Duplicate(SEXP vec, Rboolean) {
-    const auto& array = Get(vec);
-    auto size = array->length();
+      // create a standard R vector
+      SEXP copy = PROTECT(Rf_allocVector(sexp_type, array_->length()));
 
-    SEXP copy = PROTECT(Rf_allocVector(sexp_type, array->length()));
+      // copy the data from the array, through Get_region
+      Get_region(0, size, reinterpret_cast<data_type*>(DATAPTR(copy)));
 
-    memcpy(DATAPTR(copy), Dataptr_or_null(vec), size * sizeof(data_type));
+      // store as data2
+      MARK_NOT_MUTABLE(copy);
 
-    UNPROTECT(1);
-    return copy;
+      R_set_altrep_data2(alt_, copy);
+      UNPROTECT(1);
+    }
   }
 
-  static void* Dataptr(SEXP vec, Rboolean writeable) {
-    return const_cast<void*>(Dataptr_or_null(vec));
+  SEXP Duplicate(Rboolean deep) {
+    Materialize();
+    return Rf_lazy_duplicate(data2());
   }
 
-  // by definition, there are no NA
-  static int No_NA(SEXP vec) { return 1; }
+  Rboolean Inspect(int pre, int deep, int pvec,
+                   void (*inspect_subtree)(SEXP, int, int, int)) {
+    Rprintf("arrow::Array<%s, %d nulls, %s> len=%d, Array=<%p>\n",
+            array_->type()->ToString().c_str(), array_->null_count(),
+            IsMaterialized() ? "materialized" : "not materialized", array_->length(),
+            array_.get());
+    inspect_subtree(data1(), pre, deep + 1, pvec);
+    if (IsMaterialized()) {
+      inspect_subtree(data2(), pre, deep + 1, pvec);
+    }
+    return TRUE;
+  }
+
+  const void* Dataptr_or_null() {
+    if (array_->null_count() == 0) {
+      return reinterpret_cast<const void*>(const_array_data());
+    }
 
-  static void Init(R_altrep_class_t class_t, DllInfo* dll) {
-    // altrep
-    R_set_altrep_Length_method(class_t, ArrayNoNull::Length);
-    R_set_altrep_Inspect_method(class_t, ArrayNoNull::Inspect);
-    R_set_altrep_Duplicate_method(class_t, ArrayNoNull::Duplicate);
+    if (IsMaterialized()) {
+      return DATAPTR_RO(data2());
+    }
 
-    // altvec
-    R_set_altvec_Dataptr_method(class_t, ArrayNoNull::Dataptr);
-    R_set_altvec_Dataptr_or_null_method(class_t, ArrayNoNull::Dataptr_or_null);
+    return NULL;
   }
-};
 
-struct DoubleArrayNoNull {
-  static R_altrep_class_t class_t;
+  // force materialization, and then return data ptr
+  void* Dataptr(Rboolean writeable) {
+    if (array_->null_count() == 0) {
+      return reinterpret_cast<void*>(array_data());
```

Comment:
Does `writeable` mean the user wants to write to the data? If so, we should not return the original Arrow data in this case.

Reply:
I need to check that, but yeah probably

Reply:
In a way yes, but the user should not want that because the object is marked as immutable (in the R sense). Unfortunately, this is more of a guideline than something that is enforced by R. 

Accessing data read-only is somewhat new in R, and lots of internal R code would still call `DATAPTR()` even though a read)-only equivalent would be sufficient.